Add pull API

joaopapereira opened the pull request:
Introduce the new API for pull command
Enables the output as YAML for the pull command to ease the integration with other tools

Introduces the notion of a pulled image/bundle being cacheable

Fixes #390

joaopapereira committed:
Add pull API

Introduce the new API for pull command
Enables the output as YAML for the pull command to ease the integration
with other tools

Introduces the notion of a pulled image/bundle to be cacheable

Signed-off-by: Joao Pereira <joaod@vmware.com>

cppforlife commented on pkg/imgpkg/bundle/bundle.go:
```diff
@@ -31,6 +31,7 @@ type Logger interface {
 	Debugf(msg string, args ...interface{})
 	Tracef(msg string, args ...interface{})
 	Logf(msg string, args ...interface{})
+	UI() goui.UI
```
hmm, seems a bit weird for logger to carry ui. seems like we are spilling out internal implementation.

cppforlife commented on pkg/imgpkg/bundle/bundle.go:
```diff
@@ -45,11 +46,23 @@ type ImagesMetadata interface {
 }
 
+// GraphNode Node information of a Bundle
+type GraphNode struct {
```
GraphNode does not really indicate what it is. is it some type of content within a bundle? what would be a better name for this?

joaopapereira replied:
I had called it PullGraphNode before this commit but it felt too specific and we might use this structure in the future to keep some resemblance to the structure of the bundle. By structure, I am saying the bundles associated with the current bundle.

joaopapereira replied:
Since it was in the `bundle` package I assumed that `bundle.GraphNode` would have enough information. What do you think?

cppforlife commented on pkg/imgpkg/cmd/pull.go:
```diff
@@ -49,6 +58,7 @@ func NewPullCmd(o *PullOptions) *cobra.Command {
 	o.BundleRecursiveFlags.Set(cmd)
 	o.LockInputFlags.Set(cmd)
 	cmd.Flags().StringVarP(&o.OutputPath, "output", "o", "", "Output directory path")
+	cmd.Flags().StringVarP(&o.OutputType, "output-type", "", "", "Will print machine readable output on standard out: [yaml]")
```
since we have a flag called --output that takes a directory path, it's pretty confusing to have a --output-type flag that controls unrelated functionality.

joaopapereira replied:
I thought about that but we already were using output-type for the same thing in describe command. 
It would be strange to have the same behavior behind 2 different flags.

What is your suggestion?

cppforlife replied:
im not sure tbh. asking others in slack.

pivotaljohn replied:
See also: [#carvel ~ 2022-08-16](https://kubernetes.slack.com/archives/CH8KCCKA5/p1660581869764349?thread_ts=1660581817.485209&cid=CH8KCCKA5)

cppforlife commented on pkg/imgpkg/v1/pull.go:
```diff
@@ -0,0 +1,207 @@
+// BundleInfo Information related to the specific bundle
+type BundleInfo struct {
+	ImageRef      string          `json:"image"`
+	ImagesLock    *ImagesLockInfo `json:"imagesLock,omitempty"`
+	NestedBundles []BundleInfo    `json:"inner-bundles,omitempty"`
```
inconsistent casing of json names

cppforlife commented on pkg/imgpkg/v1/pull.go:
```diff
@@ -0,0 +1,207 @@
+}
+
+// InfoFromNestedBundle Given a Bundle struct it can generate the BundleInfo associated to the nested bundles
+func InfoFromNestedBundle(b *bundle.Bundle, updated bool) []BundleInfo {
```
it's not clear from the function signature what is `updated`

cppforlife replied:
do we need to make this function public?

cppforlife commented on pkg/imgpkg/v1/pull.go:
```diff
@@ -0,0 +1,207 @@
+		return st, nil
+
+	case isBundle && pullOptions.IsBundle: // Trying to pull a Bundle
+		return pullBundle(imageRef, outputPath, pullOptions, reg, false)
```
it seems like we are now making an extra registry call to check if something is a bundle... bundle.NewBundle().IsBundle and we build another copy of the bundle inside pullBundle.

cppforlife commented on pkg/imgpkg/v1/pull.go:
```diff
@@ -0,0 +1,207 @@
+}
+
+// NewBundleInfo create a BundleInfo struct from bundle information
+func NewBundleInfo(node bundle.GraphNode, updated bool) BundleInfo {
```
do we need to make this function public?

joaopapereira committed:
Change logging througout

- Create new interfaces
- Create new loggers
- Updated names of loggers
- Clarify TTY and no TTY output

Signed-off-by: Joao Pereira <joaod@vmware.com>

joaopapereira committed:
Correct linter issues

Signed-off-by: Joao Pereira <joaod@vmware.com>

joaopapereira committed:
Remove --output-type option from pull

This commit can be reverted in order to get this feature back to the
code. We decided to, for now, remove it since we are not going to use it
in the short term.
Nevertheless the basic implementation will still be present in case we
want to reintroduce this change in the future.

Signed-off-by: Joao Pereira <joaod@vmware.com>

cppforlife commented:
lgtm
